src/capi/mod.rs

```diff
@@ -26,33 +26,38 @@ fn virtblocks_util_build_file_name(file_name: *mut *mut c_char,
 
 #[no_mangle]
 pub extern "C"
-fn virtblocks_devices_memballoon_new() -> Box<devices::Memballoon> {
-    Box::new(devices::Memballoon::new())
+fn virtblocks_devices_memballoon_new() -> *mut devices::Memballoon {
+    Box::into_raw(Box::new(devices::Memballoon::new()))
 }
 
 #[no_mangle]
 pub extern "C"
-fn virtblocks_devices_memballoon_free(_memballoon: devices::Memballoon) {}
+fn virtblocks_devices_memballoon_free(c_memballoon: *mut devices::Memballoon) {
+    unsafe { Box::from_raw(c_memballoon); }
+}
 
 #[no_mangle]
 pub extern "C"
-fn virtblocks_devices_memballoon_set_model(memballoon: &mut devices::Memballoon,
+fn virtblocks_devices_memballoon_set_model(c_memballoon: *mut devices::Memballoon,
                                            model: devices::MemballoonModel)
 {
-    memballoon.set_model(model)
+    let rust_memballoon = unsafe { &mut *c_memballoon };
```

Comment:
My understanding here is that the unsafe here is due to the pointer de-reference, so there should be a non-null check. Even though we can just panic, I think it is safer than maybe-UB.

Reply:
> My understanding here is that the unsafe here is due to the pointer de-reference, so there should be a non-null check. Even though we can just panic, I think it is safer than maybe-UB.

Yeah, we definitely need to do that. `assert!()`ing as done [here](http://jakegoulding.com/rust-ffi-omnibus/objects/) should work.